grammar/SubstraitType.g4

```diff
@@ -0,0 +1,71 @@
+grammar SubstraitType;
+
+import SubstraitLexer;
+
+startRule: expr EOF;
+
+typeStatement: typeDef EOF;
+
+scalarType
+  : Boolean                 #boolean
+  | I8                      #i8
+  | I16                     #i16
+  | I32                     #i32
+  | I64                     #i64
+  | FP32                    #fp32
+  | FP64                    #fp64
+  | String                  #string
+  | Binary                  #binary
+  | Timestamp               #timestamp
+  | Timestamp_TZ            #timestampTz
+  | Date                    #date
+  | Time                    #time
+  | Interval_Year           #intervalYear
+  | UUID                    #uuid
+  ;
+
+parameterizedType
+  : FixedChar isnull=QMark? Lt length=numericParameter Gt                                   #fixedChar
+  | VarChar isnull=QMark? Lt length=numericParameter Gt                                     #varChar
+  | FixedBinary isnull=QMark? Lt length=numericParameter Gt                                 #fixedBinary
+  | Decimal isnull=QMark? Lt precision=numericParameter Comma scale=numericParameter Gt     #decimal
+  | Interval_Day isnull=QMark? Lt precision=numericParameter Gt                             #precisionIntervalDay
+  | Precision_Timestamp isnull=QMark? Lt precision=numericParameter Gt                      #precisionTimestamp
+  | Precision_Timestamp_TZ isnull=QMark? Lt precision=numericParameter Gt                   #precisionTimestampTZ
+  | Struct isnull=QMark? Lt expr (Comma expr)* Gt                                           #struct
+  | NStruct isnull=QMark? Lt Identifier expr (Comma Identifier expr)* Gt                    #nStruct
+  | List isnull=QMark? Lt expr Gt                                                           #list
+  | Map isnull=QMark? Lt key=expr Comma value=expr Gt                                       #map
+  | UserDefined Identifier isnull=QMark? (Lt expr (Comma expr)* Gt)?                        #userDefined
+  ;
+
+numericParameter
+  : Number              #numericLiteral
+  | Identifier          #numericParameterName
+  | expr                #numericExpression
+  ;
+
+anyType
+    : Any isnull=QMark?
+    | AnyVar isnull=QMark?
+    ;
+
+typeDef
+  : scalarType isnull=QMark?
+  | parameterizedType
+  | anyType
+  ;
+
+expr
+  : OParen expr CParen                                                                      #ParenExpression
+  | Identifier Eq expr Newline+ (Identifier Eq expr Newline+)* finalType=typeDef Newline*   #MultilineDefinition
+  | typeDef                                                                                 #TypeLiteral
+  | number=Number                                                                           #LiteralNumber
+  | identifier=Identifier isnull=QMark?                                                     #TypeParam
```

Comment:
Is calling this a TypeParam correct? The value might be either a type or numeric, no?

Reply:
Are you sure that's accurate? I think `varchar<L1>` will fall under TypeLiteral and as a result L1 will be covered by `numericParameter` (typeDef -> parameterizedType -> varchar) which makes sense because L1 can only ever be a numeric, never a type.

#TypeParam should cover any time parameters are used outside of the parameterized type context. for example in the "add" function:
```
init_scale = max(S1,S2)
init_prec = init_scale + max(P1 - S1, P2 - S2) + 1
min_scale = min(init_scale, 6)
delta = init_prec - 38
prec = min(init_prec, 38)
scale_after_borrow = max(init_scale - delta, min_scale)
scale = init_prec > 38 ? scale_after_borrow : init_scale
DECIMAL<prec, scale>
```

Every time a parameter is used (except for the last expression) params will be resolved to `TypeParam`, while `prec` and `scale` at the end will be resolved to `NumericParameterName` instead.

All I'm saying is that when we have two expressions named `NumericParameterName` and `TypeParam`, it makes you think that one can only hold numerics and the other can only hold types, when in fact `TypeParam` can hold both.

I'd suggest using more generic `ParameterName` instead of `TypeParam` to make that distinction less confusing, plus make the name more consistent with `NumericParameterName`.

Reply:
You are Right, for varchar `numericParameter` covers `L1` and `10`. I was confusing it with parameters of UserDefinedType.

I will change the name to `ParameterName`

Reply:
thanks. let me take the discussion a bit further though. I was thinking, is there a point in having a separate `numericParameter` token? If the goal is to somehow limit what is or is not applicable in that context, I don't think that can be achieved anyway. `numericParameter` can contain an `expr` which may or may not be numeric. On the flip side, both number (literal) and identifier (parameter) can be expressed by expr anyway. Can't we have exprs instead of numericParameters inside parameterizedType?:

```
FixedChar isnull=QMark? Lt length=expr Gt                                   #fixedChar
...
```

Reply:
I see the duplication of rules, but I also see current grammar is helpful in parser code to create different type of objects based on the context. 

I haven't yet worked on the type derivation, but this is the grammar that is used in other substrait repos, so there must be a good reason.

Reply:
I just found it redundant when I was doing an implementation in python. I'm fine with leaving it as is as well, though.